calico/openstack/mech_calico.py

```diff
@@ -196,14 +205,163 @@ def _init_state(self):
+                                             wait=True,
+                                             waitIndex = self.next_etcd_index,
+                                             recursive=True)
+
+                # Before status is passed to Neutron db, check whether we
+                # are still master (if not, we loop until we become one)
+                # and whether epoch is as expected (if not, thread exits).
+                if self._epoch != expected_epoch or not self.transport.is_master:
+                    break
+
+                self._handle_status_update(response)
+                # Since we're polling on a subtree, we can't just
+                # increment the index, we have to look at the
+                # modifiedIndex.
+                self.next_etcd_index = max(self.next_etcd_index,
+                                           int(response.modifiedIndex)) + 1
+        except (ReadTimeoutError, SocketTimeout) as e:
+            # This is expected when we're doing a poll and nothing
+            # happened. socket timeout doesn't seem to be caught by
+            # urllib3 1.7.1. Simply reconnect.
+            LOG.debug("Read from etcd timed out (%r), retrying.", e)
+            self._client = etcd.Client(host=cfg.CONF.calico.etcd_host,
+                                       port=cfg.CONF.calico.etcd_port)
+        except (etcd.EtcdClusterIdChanged, etcd.EtcdEventIndexCleared) as e:
+            # Complete resync is needed.
+            LOG.warning("Out of sync with etcd (%r).  Reconnecting "
+                     "for full sync.", e)
+            try:
+                # Connect the client
+                self._client = etcd.Client(host=cfg.CONF.calico.etcd_host,
+                                           port=cfg.CONF.calico.etcd_port)
+                context = ctx.get_admin_context()
+                # Resync endpoints.
+                self.resync_endpoints(context)
+                # Resync profiles.
+                self.resync_profiles(context)
+                # Now, set the config flags.
+                self.transport.provide_felix_config()
+            except Exception:
+                LOG.exception("Error in full resync.")
+        except Exception as e:
+            ## What other exceptions do we expect?
+            ## Is full resync needed on unexpected exception?
+            LOG.info("Handling status update failed. ({})".format(e))
+            self._client = etcd.Client(host=cfg.CONF.calico.etcd_host,
+                                       port=cfg.CONF.calico.etcd_port)
+            eventlet.sleep(10)
+
+    def _handle_status_update(self, response):
+        """
+        Handle updates of felix instances written to etcd. Pass status
+        updates to neutron database.
+
+        :param response: update written to etcd by one of felix instances
+        """
+        # Only get info from uptime keys - since those have ttl, we know
+        # if key was updated or created.
```

Comment:
I'm not keen on the magic number 7 here - please define a constant, or use a more explicit regex to show your assumptions about the key structure.